server/src/main/java/org/opensearch/index/query/CombinedFieldQueryBuilder.java

```diff
@@ -0,0 +1,335 @@
+     */
+    public CombinedFieldQueryBuilder field(String field, float weight) {
+        if (Strings.isEmpty(field)) {
+            throw new IllegalArgumentException("supplied field is null or empty.");
+        }
+        checkValidWeight(weight);
+        this.fieldWeights.put(field, weight);
+        return this;
+    }
+
+    private static void checkValidWeight(float weight) {
+        if (Float.compare(weight, 1f) < 0) {
+            throw new IllegalArgumentException(("weights must be greater or equal to 1 in [" + NAME + "]"));
+        }
+    }
+
+    /**
+     * Add several fields to run the query against with a specific weight.
+     */
+    public CombinedFieldQueryBuilder fields(Map<String, Float> fields) {
+        for (Map.Entry<String, Float> fieldWeight : fields.entrySet()) {
+            field(fieldWeight.getKey(), fieldWeight.getValue());
+        }
+        return this;
+    }
+
+    public Map<String, Float> fields() {
+        return fieldWeights;
+    }
+
+    /**
+     * Explicitly set the analyzer to use. Defaults to use explicit mapping config for the field, or, if not
+     * set, the default search analyzer.
+     */
+    public CombinedFieldQueryBuilder analyzer(String analyzer) {
+        this.analyzer = analyzer;
+        return this;
+    }
+
+    public String analyzer() {
+        return analyzer;
+    }
+
+    public CombinedFieldQueryBuilder zeroTermsQuery(MatchQuery.ZeroTermsQuery zeroTermsQuery) {
+        if (zeroTermsQuery == null) {
+            throw new IllegalArgumentException("[" + NAME + "] requires zero terms query to be non-null");
+        }
+        this.zeroTermsQuery = zeroTermsQuery;
+        return this;
+    }
+
+    public MatchQuery.ZeroTermsQuery zeroTermsQuery() {
+        return zeroTermsQuery;
+    }
+
+    static boolean isTextField(MappedFieldType fieldType) {
+        if (fieldType == null) {
+            return false;
+        }
+        return fieldType.getTextSearchInfo() != TextSearchInfo.NONE && fieldType.getTextSearchInfo() != TextSearchInfo.SIMPLE_MATCH_ONLY;
+    }
+
+    @Override
+    protected Query doToQuery(QueryShardContext context) throws IOException {
+        boolean hasTextField = fieldWeights.keySet().stream().anyMatch(k -> isTextField(context.fieldMapper(k)));
+        if (hasTextField == false) {
+            return Queries.newUnmappedFieldsQuery(fieldWeights.keySet());
+        }
+        Analyzer explicitAnalyzer = null;
+        if (analyzer != null) {
+            explicitAnalyzer = context.getMapperService().getIndexAnalyzers().get(analyzer);
+            if (explicitAnalyzer == null) {
+                throw new IllegalArgumentException("No analyzer found for [" + analyzer + "]");
+            }
+        }
+
+        CombinedFieldQuery.Builder builder = new CombinedFieldQuery.Builder();
+        boolean hasTerms = false;
+        for (Map.Entry<String, Float> fieldWeight : fieldWeights.entrySet()) {
```

Comment:
I read you remark regarding empty fields and non-text fields, may be better option is to reject a query and return the error to the user instead of silently skipping fields over and returning no matches? Reasoning: it is going to be difficult to troubleshoot why certain result set is returned without understanding that certain field just are not taken into account. What do you think?

Reply:
That's a really good point. I like the idea of failing rather than mysteriously pretending invalid input doesn't exist.

I did the "silently skipping" thing to try to be consistent with [MultiMatchQuery](https://github.com/opensearch-project/OpenSearch/blob/main/server/src/main/java/org/opensearch/index/search/MultiMatchQuery.java#L124-L126), but we probably don't need to repeat that approach.

In particular, I think we buy ourselves more flexibility by failing initially. If someone has a compelling reason why we shouldn't fail, we can stop failing. It's harder to start failing on queries that were previously valid (since some people may be relying on the lenient behavior).

I'll update in a subsequent commit.

Reply:
:bow: